lib/membrane_rtmp_plugin/rtmp/source.ex

```diff
@@ -58,10 +66,40 @@ defmodule Membrane.RTMP.Source do
 
   @impl true
   def handle_demand(pad, _size, _unit, _ctx, state) do
-    case Native.read_frame(state.native) do
-      {:ok, type, frame} ->
+    {actions, state} = maybe_provide_frame(pad, state)
+    send(state.provider, :get_frame)
+    {{:ok, actions}, state}
+  end
+
+  defp maybe_provide_frame(type, state) do
+    buffer = get_in(state, [:buffers, type])
+
+    {actions, queue} =
+      case Qex.pop(buffer) do
+        {:empty, queue} -> {[], queue}
+        {{:value, buffer}, queue} -> {[buffer: {type, buffer}, redemand: type], queue}
+      end
+
+    state = put_in(state, [:buffers, type], queue)
+    {actions, state}
+  end
+
+  @impl true
+  def handle_other({:result, result}, ctx, state) when ctx.playback_state == :playing do
```

Comment:
maybe `{:frame_provider, message}` and then
`case message do`

or split this into multiple function clauses